File: app/src/main/java/com/wcaokaze/android/twiqua/activity/main/widget/columnlayout/ColumnLayout.java
/*
 * Copyright 2020 wcaokaze
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wcaokaze.android.twiqua.activity.main.widget.columnlayout;

import android.content.Context;
import android.view.MotionEvent;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

public final class ColumnLayout extends FrameLayout {
   private int mVisibleColumnCount = 1;
   private int mColumnMargin = 0;
   private int mPadding = 0;

   @Nullable private ColumnLayoutAdapter mAdapter = null;
   @Nullable private ColumnLayoutManager mLayoutManager = null;
   @Nullable private ColumnLayoutGestureDetector<?> mGestureDetector = null;

   public ColumnLayout(final Context context) {
      super(context);
   }

   @Nullable
   public final ColumnLayoutAdapter getAdapter() {
      return mAdapter;
   }

   public final void setAdapter(@Nullable final ColumnLayoutAdapter adapter) {
      mAdapter = adapter;
      relayout();
   }

   public final int getColumnMargin() {
      return mColumnMargin;
   }

   public final void setColumnMargin(final int columnMargin) {
      mColumnMargin = columnMargin;
      relayout();
   }

   @Nullable
   public final ColumnLayoutGestureDetector<?> getGestureDetector() {
      return mGestureDetector;
   }

   public final void setGestureDetector
         (@Nullable final ColumnLayoutGestureDetector<?> gestureDetector)
   {
      mGestureDetector = gestureDetector;
   }

   @Nullable
   public final ColumnLayoutManager getLayoutManager() {
      return mLayoutManager;
   }

   public final void setLayoutManager(@Nullable final ColumnLayoutManager layoutManager) {
      if (mLayoutManager != null) {
         mLayoutManager.onDetachedFromColumnLayout(this);
      }

      mLayoutManager = layoutManager;

      if (layoutManager != null) {
         layoutManager.onAttachedToColumnLayout(this);
      }

      relayout();
   }

   public final int getPadding() {
      return mPadding;
   }

   public final void setPadding(final int padding) {
      mPadding = padding;
      relayout();
   }

   public final int getVisibleColumnCount() {
      return mVisibleColumnCount;
   }

   public final void setVisibleColumnCount(final int visibleColumnCount) {
      mVisibleColumnCount = visibleColumnCount;
      relayout();
   }

   @Override
   public final boolean onInterceptTouchEvent(final MotionEvent ev) {
      final ColumnLayoutGestureDetector<?> gestureDetector = mGestureDetector;
      if (gestureDetector == null) { return false; }
      return gestureDetector.onInterceptTouchEvent(this, ev);
   }

   @Override
   public final boolean onTouchEvent(final MotionEvent event) {
      final ColumnLayoutGestureDetector<?> gestureDetector = mGestureDetector;
      if (gestureDetector == null) { return false; }
      return gestureDetector.onTouchEvent(this, event);
   }

   private void relayout() {
      final ColumnLayoutManager layoutManager = mLayoutManager;
      if (layoutManager != null) {
         layoutManager.relayout(this);
      }
   }

   @Override
   protected final void onSizeChanged(final int w, final int h,
                                      final int oldw, final int oldh)
   {
      super.onSizeChanged(w, h, oldw, oldh);

      post(new Runnable() {
         @Override
         public void run() {
            relayout();
         }
      });
   }
}

File: app/src/main/java/com/wcaokaze/android/twiqua/activity/main/widget/_ColumnLayout.kt
/*
 * Copyright 2020 wcaokaze
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("UNUSED")
package com.wcaokaze.android.twiqua.activity.main.widget

import android.content.Context
import koshian.*
import kotlin.contracts.*

object ColumnLayoutConstructor : KoshianViewConstructor<ColumnLayout> {
   override fun instantiate(context: Context?) = ColumnLayout(context)
}

/**
 * creates a new ColumnLayout and adds it into this ViewGroup.
 */
@ExperimentalContracts
@Suppress("FunctionName")
inline fun <L> CreatorParent<L>.ColumnLayout(
   creatorAction: ViewCreator<ColumnLayout, L>.() -> Unit
): ColumnLayout {
   contract { callsInPlace(creatorAction, InvocationKind.EXACTLY_ONCE) }
   return create(ColumnLayoutConstructor, creatorAction)
}

/**
 * creates a new ColumnLayout with name, and adds it into this ViewGroup.
 *
 * The name can be referenced in [applyKoshian]
 */
@ExperimentalContracts
@Suppress("FunctionName")
inline fun <L> CreatorParent<L>.ColumnLayout(
   name: String,
   creatorAction: ViewCreator<ColumnLayout, L>.() -> Unit
): ColumnLayout {
   contract { callsInPlace(creatorAction, InvocationKind.EXACTLY_ONCE) }
   return create(name, ColumnLayoutConstructor, creatorAction)
}

/**
 * If the next View is a ColumnLayout, applies Koshian to it.
 *
 * Otherwise, creates a new ColumnLayout and inserts it to the current position.
 *
 * @see applyKoshian
 */
@Suppress("FunctionName")
inline fun <L, S : KoshianStyle>
      ApplierParent<L, S>.ColumnLayout(
         applierAction: ViewApplier<ColumnLayout, L, S>.() -> Unit
      )
{
   apply(ColumnLayoutConstructor, applierAction)
}

/**
 * If the next View is a ColumnLayout, applies Koshian to it.
 *
 * Otherwise, creates a new ColumnLayout and inserts it to the current position.
 *
 * @see applyKoshian
 */
@Suppress("FunctionName")
inline fun <L, S : KoshianStyle>
      ApplierParent<L, S>.ColumnLayout(
         styleElement: KoshianStyle.StyleElement<ColumnLayout>,
         applierAction: ViewApplier<ColumnLayout, L, S>.() -> Unit
      )
{
   apply(ColumnLayoutConstructor, styleElement, applierAction)
}

/**
 * Applies Koshian to all ColumnLayouts that are named the specified in this ViewGroup.
 * If there are no ColumnLayouts named the specified, do nothing.
 *
 * @see applyKoshian
 */
@Suppress("FunctionName")
inline fun <L, S : KoshianStyle>
      ApplierParent<L, S>.ColumnLayout(
         name: String,
         applierAction: ViewApplier<ColumnLayout, L, S>.() -> Unit
      )
{
   apply(name, applierAction)
}

/**
 * Applies Koshian to all ColumnLayouts that are named the specified in this ViewGroup.
 * If there are no ColumnLayouts named the specified, do nothing.
 *
 * @see applyKoshian
 */
@Suppress("FunctionName")
inline fun <L, S : KoshianStyle>
      ApplierParent<L, S>.ColumnLayout(
         name: String,
         styleElement: KoshianStyle.StyleElement<ColumnLayout>,
         applierAction: ViewApplier<ColumnLayout, L, S>.() -> Unit
      )
{
   apply(name, styleElement, applierAction)
}

/**
 * registers a style applier function into this [KoshianStyle].
 *
 * Styles can be applied via [applyKoshian]
 */
@Suppress("FunctionName")
inline fun KoshianStyle.ColumnLayout(
   crossinline styleAction: ViewStyle<ColumnLayout>.() -> Unit
): KoshianStyle.StyleElement<ColumnLayout> {
   return createStyleElement(styleAction)
}
